src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -601,32 +911,247 @@ private static PartialCaseDecision RemoveEvaluation(PartialCaseDecision c, Bound
                 Index: c.Index,
                 Syntax: c.Syntax,
                 Decisions: c.Decisions.WhereAsArray(d => !(d is BoundDagEvaluation e2) || e2 != e),
-                Bindings: c.Bindings, WhereClause: c.WhereClause, CaseLabel: c.CaseLabel);
+                Bindings: c.Bindings, WhenClause: c.WhenClause, CaseLabel: c.CaseLabel);
+        }
+
+        /// <summary>
+        /// The state at a given node of the decision acyclic graph. This is used during computation of the state machine,
+        /// and contains a representation of the meaning of the state.
+        /// </summary>
+        private class DagState
+        {
+            public readonly ImmutableArray<PartialCaseDecision> Cases;
+
+            public DagState(ImmutableArray<PartialCaseDecision> cases)
+            {
+                this.Cases = cases;
+            }
+
+            public static implicit operator DagState(ImmutableArray<PartialCaseDecision> cases) => new DagState(cases);
+
+            // We only compute the dag states for the branches after we de-dup this DagState itself.
+            // If all that remains is the `when` clauses, SelectedDecision is left `null` and the
+            // FalseBranch field is populated with the successor on failure of the when clause (if one exists).
+            public BoundDagDecision SelectedDecision;
+            public DagState TrueBranch, FalseBranch;
+
+            public ImmutableArray<DagState> TopologicallySortedReachableStates
```

Comment:
>TopologicallySortedReachableStates [](start = 44, length = 34)

Consider changing this property to a function since the value is relatively expensive to calculate and is calculated each time.

Reply:
It is now cached in [https://github.com/dotnet/roslyn/pull/25662](https://github.com/dotnet/roslyn/pull/25662)

---
In reply to: [175851167](https://github.com/dotnet/roslyn/pull/25544#discussion_r175851167) [](ancestors = 175851167)